### PR TITLE
fix: reuse

### LINKED ISSUE #14: Event Reusable
## Problem Definition
**Is your feature request related to a problem? Please describe.**

1. Currently, the Admin Portal and User Portal login screens duplicate code and styling that can get out of sync.
2. This makes maintenance difficult and has historically led to UI/UX inconsistencies between portals.
3. Though the Admin and User Portal share components for events it is not consistent with our reusable components guide.
   - https://docs-admin.talawa.io/docs/developer-resources/reusable-components

**Describe the solution you'd like**

1. Use reusable components for these elements
2. **Refactor the code:**
   
   1. to match that of the Figma design AND have a consistent UI/UX
      
      * https://docs.talawa.io/docs/design/ux-design-system#talawa-admin---figma-designs
   2. exclusively use shared components and appropriate props for both the Admin and User portal. When in doubt, follow the guidance here:
      
      * https://docs-admin.talawa.io/docs/developer-resources/reusable-components
3. Ensure that existing features are maintained in each screen
4. Update the test suite to achieve 100% code coverage for your patch.
5. Ensure all tests submitted are functional and valid
6. Write tests to ensure this feature doesn't break.

**Additional context**

Image

### LINKED ISSUE #15: User Portal
**Describe the bug**
User Portal - There is no ability to set or modify the interval of a recurring event

**To Reproduce**
1. Scenario 1 - Creating a recurring event 
   1. Login
   2. Select an organization
   3. Click on the event screen
   4. Click on the add event button
   5. Select recurring
   6. There is no option for the interval   
      1. Daily
      2. Monthly
      3. Weekly etc.
2. Scenario 2 - Editing a recurring event  
   1. Login
   2. Select an organization
   3. Click on the event screen
   4. Select a recurring event
   5. Visit the event dashboard
   6. There is no option for the interval:     
      1. Daily
      2. Monthly
      3. Weekly etc.
   7. There is no option to:   
      1. see or edit the event interval.

**Expected behavior**
The operation needs to match that of Admin Portal

### PR DESCRIPTION
**Pull request by rawadhossain**
**What kind of change does this PR introduce?**

**Issue Number:**

Fixes #14 
Fixes #15

**Snapshots/Videos:**

**If relevant, did you update the documentation?**

**Summary**
1. **Extract Recurrence Picker as Shared Component**
Created src/shared-components/EventRecurrencePicker/:
   - `EventRecurrencePicker.tsx` - The shared recurrence picker component
   - `EventRecurrencePicker.spec.tsx` - Unit tests (9 tests, all passing)
   - `EventRecurrencePickerMocks.ts` - Mock data for tests
Created `src/types/EventRecurrencePicker/interface.ts` 

2. **Refactor Existing Usages**
Updated these files to use the shared component:
   - `src/screens/OrganizationEvents/CreateEventModal.tsx` - Admin event creation
   - `src/components/EventListCard/Modal/Preview/EventListCardPreviewModal.tsx` - Event editing modal
3. **Integrate into User Portal Event Creation** 
Updated `src/screens/UserPortal/Events/Events.tsx`:
   - Replaced simple `isRecurring` toggle with `EventRecurrencePicker`  
   - Added `recurrence` state using `InterfaceRecurrenceRule | null`
   - Updated `createEvent` mutation to include recurrence payload with validation
 
**Does this PR introduce a breaking change?**

### CodeRabbit AI Review
- [ ] I have reviewed and addressed all critical issues flagged by CodeRabbit AI
- [ ] I have implemented or provided justification for each non-critical suggestion
- [ ] I have documented my reasoning in the PR comments where CodeRabbit AI suggestions were not implemented

### Test Coverage
- [ ] I have written tests for all new changes/features
- [ ] I have verified that test coverage meets or exceeds 95%
- [ ] I have run the test suite locally and all tests pass

**Other information**

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

## Summary by CodeRabbit

* **New Features**
  * Added a unified Event Recurrence Picker for creating and previewing events with standard options, custom flow, end-date handling, defaults, and disabled state.

* **Refactor**
  * Simplified recurrence wiring by replacing legacy dropdown/modal plumbing and removing obsolete preview modal recurrence props.

* **Tests**
  * Added and updated tests covering picker options, labels, custom modal flow, end-date updates, validation, and event-creation integration.

* **Localization**
  * Added recurrence-related translations across locales (Does not repeat, Daily, Weekly, Monthly, Annually, Every weekday, Custom...).

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: src/shared-components/EventRecurrencePicker/EventRecurrencePicker.tsx [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+    'June',
+    'July',
+    'August',
+    'September',
+    'October',
+    'November',
+    'December',
+  ];
+  return months[monthIndex];
+};
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/shared-components/EventRecurrencePicker/EventRecurrencePicker.tsx]: @coderabbitai so what do you suggest? give me the replacement codes for day and month

File: src/shared-components/EventRecurrencePicker/EventRecurrencePicker.tsx [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+          byDay: ['MO', 'TU', 'WE', 'TH', 'FR'] as WeekDays[],
+          never: true,
+        },
+      },
+      {
+        label: 'Custom...',
+        value: 'custom',
+      },
+    ];
+  };
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/shared-components/EventRecurrencePicker/EventRecurrencePicker.tsx:185]: Given the small and inexpensive computation here, I opted to keep it simple and avoid premature memoization.

**Comment by rawadhossain** [comment]
@coderabbitai apart from the requested trivial changes and insufficient patch coverage, is everything in the PR correct? Have I addressed the changes as required for both of the issues?
Do you have any recommendation what else should I do or not.

File: src/components/EventListCard/Modal/Preview/EventListCardPreviewModal.spec.tsx [hunk 3]
```diff
@@ -923,140 +918,109 @@ describe('EventListCardPreviewModal', () => {
...
+      });
 
-      // Verify the t function is passed and works correctly
-      expect(customModalProps.t).toBeDefined();
-      expect(typeof customModalProps.t).toBe('function');
-      expect(customModalProps.t('testKey')).toBe('testKey');
+      expect(
+        screen.queryByTestId('recurrenceDropdown'),
+      ).not.toBeInTheDocument();
     });
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/components/EventListCard/Modal/Preview/EventListCardPreviewModal.spec.tsx:1068]: I agree there is some overlap, but I’ve kept the tests explicit for clarity and to keep this PR scoped to the recurrence refactor.

**Comment by rawadhossain** [comment]
> Tip
> 
> For best results, initiate chat on the files or code changes.
> 
> 🧩 Analysis chain
> 🏁 Script executed:
> 
> ```
> #!/bin/bash
> # Check if User Portal has edit event functionality with recurrence support
> echo "=== Searching for User Portal edit event functionality ==="
> rg -n -A5 -B5 "edit.*event|update.*event" src/screens/UserPortal/Events/Events.tsx --type tsx 2>/dev/null || rg -n -A5 -B5 "edit.*event|update.*event" src/screens/UserPortal/Events/Events.tsx
> 
> echo -e "\n=== Check for UPDATE_EVENT mutation in User Portal ==="
> rg -n "UPDATE_EVENT|updateEvent" src/screens/UserPortal/Events/ --type tsx 2>/dev/null || rg -n "UPDATE_EVENT|updateEvent" src/screens/UserPortal/Events/
> 
> echo -e "\n=== List all files in User Portal Events directory ==="
> fd . src/screens/UserPortal/Events/ --type f
> 
> echo -e "\n=== Check recurrence validation utilities ==="
> cat src/utils/recurrenceUtils/validation.ts 2>/dev/null || echo "File not found"
> 
> echo -e "\n=== Check for any TODOs or FIXMEs in the new component ==="
> rg -n "TODO|FIXME|XXX|HACK" src/shared-components/EventRecurrencePicker/EventRecurrencePicker.tsx
> 
> echo -e "\n=== Check test coverage configuration ==="
> cat vitest.config.ts 2>/dev/null | head -50 || cat jest.config.js 2>/dev/null | head -50 || echo "No test config found"
> ```
> 
> Length of output: 1757
> 
> 🏁 Script executed:
> 
> ```
> #!/bin/bash
> # Examine the EventRecurrencePicker implementation for potential issues
> echo "=== Check EventRecurrencePicker implementation ==="
> cat src/shared-components/EventRecurrencePicker/EventRecurrencePicker.tsx | head -100
> 
> echo -e "\n=== Check if there are any accessibility concerns ==="
> rg -n "aria-|role=|alt=" src/shared-components/EventRecurrencePicker/EventRecurrencePicker.tsx
> 
> echo -e "\n=== Check User Portal Events.tsx recurrence integration ==="
> cat src/screens/UserPortal/Events/Events.tsx | grep -A30 "EventRecurrencePicker\|recurrence"
> ```
> 
> Length of output: 10766
> 
> 🏁 Script executed:
> 
> ```
> #!/bin/bash
> # Check for any modal or edit UI in User Portal
> echo "=== Search for edit/update modal in User Portal Events ==="
> rg -n "Modal|Dialog" src/screens/UserPortal/Events/Events.tsx -C3
> 
> echo -e "\n=== Check if there's a separate edit event component in User Portal ==="
> fd "edit" src/screens/UserPortal/Events/ --type f
> 
> echo -e "\n=== Check the structure of Admin Portal edit functionality ==="
> rg -n "EventListCardPreviewModal" src/components/EventListCard/ -l
> 
> echo -e "\n=== Check what the EventListCardPreviewModal does ==="
> cat src/components/EventListCard/Modal/Preview/EventListCardPreviewModal.tsx | grep -A10 "const.*PreviewModal\|function.*PreviewModal"
> ```
> 
> Length of output: 3879
> 
> 🏁 Script executed:
> 
> ```
> #!/bin/bash
> # Find where validation and formatting utilities are defined
> echo "=== Search for validateRecurrenceInput definition ==="
> rg -n "export.*validateRecurrenceInput|const.*validateRecurrenceInput|function.*validateRecurrenceInput" src/
> 
> echo -e "\n=== Search for formatRecurrenceForApi definition ==="
> rg -n "export.*formatRecurrenceForApi|const.*formatRecurrenceForApi|function.*formatRecurrenceForApi" src/
> 
> echo -e "\n=== Check recurrenceUtils directory structure ==="
> fd . src/utils/recurrenceUtils/ --type f
> 
> echo -e "\n=== Check if these utilities exist anywhere ==="
> rg -n "validateRecurrenceInput|formatRecurrenceForApi" src/utils/recurrenceUtils/ -C2
> 
> echo -e "\n=== Check test coverage report if available ==="
> cat coverage/coverage-summary.json 2>/dev/null | jq '.total' || echo "No coverage report found"
> ```
> 
> Length of output: 9868
> 
> 🏁 Script executed:
> 
> ```
> #!/bin/bash
> # Check for potential issues in the implementation
> echo "=== Check EventRecurrencePicker for error handling ==="
> rg -n "error|Error|catch|throw" src/shared-components/EventRecurrencePicker/EventRecurrencePicker.tsx
> 
> echo -e "\n=== Check if startDate validation exists ==="
> cat src/shared-components/EventRecurrencePicker/EventRecurrencePicker.tsx | grep -A5 -B5 "startDate" | head -30
> 
> echo -e "\n=== Check the test file for coverage gaps ==="
> cat src/shared-components/EventRecurrencePicker/EventRecurrencePicker.spec.tsx | grep "describe\|it(" | head -20
> 
> echo -e "\n=== Check if User Portal has any other event management screens ==="
> fd . src/screens/UserPortal/ --type d
> 
> echo -e "\n=== Check Admin Portal for reference on edit functionality ==="
> rg -n "isEditMode|editEvent|updateEvent" src/components/EventListCard/Modal/Preview/EventListCardPreviewModal.tsx | head -20
> ```
> 
> Length of output: 2749
> 
> @rawadhossain, great work on creating the reusable `EventRecurrencePicker` component! Let me provide a comprehensive assessment:
> 
> ## ✅ What's Done Well
> 1. **Issue [Event Reusable #14](https://github.com/rawadhossain/talawa-admin/issues/14) (Reusable Component)** - **Mostly Addressed**:
>    
>    * Successfully extracted a shared `EventRecurrencePicker` component with clean interface
>    * Refactored Admin Portal's `CreateEventModal` and `EventListCardPreviewModal` to use the shared component
>    * Eliminated code duplication across Admin Portal
>    * Added 9 passing unit tests for the new component
>    * Proper validation and formatting utilities are in place
> 2. **Code Quality**:
>    
>    * Clean component design with semantic comparison helpers (`recurrenceRulesEqual`, `arraysEqual`)
>    * Proper TypeScript interfaces and type safety
>    * Good documentation and comments
>    * Integration of validation (`validateRecurrenceInput`) and formatting (`formatRecurrenceForApi`)
> 
> ## ❌ Critical Gap: Issue #15 Not Fully Addressed
> **Issue #15 explicitly states**: _"Ensure both creating AND editing recurring events expose interval selection"_
> 
> Based on my analysis:
> 
> * ✅ **User Portal CREATE**: Successfully implemented with `EventRecurrencePicker`
> * ❌ **User Portal EDIT**: **Missing entirely** - `src/screens/UserPortal/Events/Events.tsx` has NO edit/update event functionality
> 
> The User Portal only has a `createEventModal` with no corresponding edit modal or `UPDATE_EVENT` mutation. To fully address Issue #15, you need to:
> 
> 1. Add edit event functionality to User Portal (similar to Admin Portal's `EventListCardPreviewModal`)
> 2. Integrate `EventRecurrencePicker` into the edit flow
> 3. Handle updating existing recurring events with new recurrence rules
> 
> ## 📊 Test Coverage Concerns
> You mentioned "insufficient patch coverage." Here are likely gaps:
> 
> 1. **EventRecurrencePicker edge cases**:
>    
>    * Invalid date handling
>    * Timezone edge cases
>    * Disabled state interactions
>    * Error scenarios
> 2. **Integration tests**:
>    
>    * User Portal event creation with various recurrence patterns
>    * End-to-end flow validation
>    * Recurrence validation error handling
> 3. **Admin Portal refactored components**:
>    
>    * Ensure coverage wasn't reduced after refactoring
>    * Test recurrence persistence and retrieval
> 
> ## 🔍 Other Recommendations
> ### Accessibility
> Your `EventRecurrencePicker` has **no accessibility attributes**. Add:
> 
> ```tsx
> <Dropdown.Toggle
>   aria-label="Select event recurrence pattern"
>   aria-describedby="recurrence-description"
> >
> ```
> 
> ### Error Handling
> Consider adding error boundaries and null-safe guards:
> 
> ```tsx
> const eventDate = startDate ? new Date(startDate) : new Date();
> if (isNaN(eventDate.getTime())) {
>   // Handle invalid date
> }
> ```
> 
> ### Missing User Portal Edit Implementation
> Check if there's a separate component for viewing User Portal events (like Admin's EventListCard). You may need to:
> 
> 1. Create an edit modal for User Portal events
> 2. Add UPDATE_EVENT mutation
> 3. Integrate EventRecurrencePicker for editing
> 
> ## 📋 Action Items to Complete the PR
> 1. **Critical**: Implement User Portal edit event functionality with recurrence support
> 2. Improve test coverage to meet 100% requirement (Issue [Event Reusable #14](https://github.com/rawadhossain/talawa-admin/issues/14))
> 3. Add accessibility attributes to EventRecurrencePicker
> 4. Consider adding integration/E2E tests for full event creation/editing flow
> 5. Update PR description to clearly note if edit functionality is deferred to a separate PR
> 
> Would you like me to help identify where to add the User Portal edit functionality, or would you prefer to create a follow-up issue for it?

@coderabbitai is it really necessary to add accessibility attributes in Dropdown EventRecurrencePicker? does adding accessibility attributes match the convention of the codebase?

File: src/screens/UserPortal/Events/Events.spec.tsx [hunk 4]
```diff
@@ -978,19 +978,65 @@ describe('Testing Events Screen [User Portal]', () => {
...
+
+    await userEvent.click(screen.getByTestId('recurrenceOption-1'));
+
+    // Should now show Daily
+    await waitFor(() => {
+      expect(screen.getByTestId('recurrenceDropdown')).toHaveTextContent(
+        'Daily',
+      );
+    });
+  });
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/screens/UserPortal/Events/Events.spec.tsx:1085]: Following existing test patterns in the codebase to keep consistency. Can migrate tests to i18n-based assertions in a separate cleanup PR if desired.

File: src/screens/UserPortal/Events/Events.spec.tsx [hunk 5]
```diff
@@ -1475,4 +1521,99 @@ describe('Testing Events Screen [User Portal]', () => {
...
+
+    await waitFor(() => {
+      expect(screen.getByTestId('recurrenceDropdown')).toHaveTextContent(
+        'Daily',
+      );
+    });
+
+    await userEvent.click(screen.getByTestId('createEventBtn'));
+    await wait(500);
+  });
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/screens/UserPortal/Events/Events.spec.tsx]: Check commit feef2dad3d124767d24c3c6831e3d51c36978424 approach changed for Events.spec.tsx.
This approach:
 1.  meaningful functionality (the original issue was about adding recurrence selection to User Portal)
2. Has real assertions that would catch regressions
3. Avoids the fragile date-matching issues with Apollo mocks
4. The mutation itself is already tested in other tests

File: src/screens/UserPortal/Events/Events.spec.tsx [hunk 6]
```diff
@@ -1475,4 +1521,90 @@ describe('Testing Events Screen [User Portal]', () => {
...
+    });
+
+    await userEvent.click(screen.getByTestId('recurrenceOption-0'));
+
+    await waitFor(() => {
+      expect(screen.getByTestId('recurrenceDropdown')).toHaveTextContent(
+        'Does not repeat',
+      );
+    });
+  });
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/screens/UserPortal/Events/Events.spec.tsx:1656]: Check latest commit, tried different approach

File: src/components/EventListCard/Modal/Preview/EventListCardPreviewModal.spec.tsx [hunk 7]
```diff
@@ -602,18 +629,55 @@ describe('EventListCardPreviewModal', () => {
...
+        screen.getByTestId('customRecurrenceModalSetEndDateBtn'),
+      ).toBeInTheDocument();
+    });
+
+    await userEvent.click(
+      screen.getByTestId('customRecurrenceModalSetEndDateBtn'),
+    );
+
+    expect(mockSetEventEndDate).toHaveBeenCalledWith(new Date('2024-02-02'));
   });
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/components/EventListCard/Modal/Preview/EventListCardPreviewModal.spec.tsx:681]: The overlap here is intentional — one test exercises the full integration flow (UI → custom modal → callback), while the simpler one guards the logic path directly. This gives coverage at different layers and helps catch regressions in either wiring or logic.

I’d prefer to keep both for now since they’re lightweight and serve distinct purposes, but I’m happy to consolidate in a follow-up cleanup PR if desired.

File: src/components/EventListCard/Modal/Preview/EventListCardPreviewModal.spec.tsx [hunk 8]
```diff
@@ -923,140 +986,176 @@ describe('EventListCardPreviewModal', () => {
...
           ...mockEventListCardProps,
           isRecurringEventTemplate: true,
-          recurrenceDescription: 'Custom Rule',
         },
         recurrence: null,
       });
-      expect(screen.getByText('Custom Rule')).toBeInTheDocument();
+      expect(screen.getByTestId('recurrenceDropdown')).toBeInTheDocument();
     });
+  });
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/components/EventListCard/Modal/Preview/EventListCardPreviewModal.spec.tsx:999]: I kept the test here since it was added while covering the recurrence label behavior, but I agree it could live in a more general rendering/integration section. To avoid restructuring in this PR (now that tests and coverage are stable), I’d prefer to leave it as is and can move it in a follow-up PR if that’s preferred.

File: src/screens/UserPortal/Events/Events.spec.tsx [hunk 9]
```diff
@@ -41,6 +46,15 @@ const { mockToast, mockUseParams } = vi.hoisted(() => ({
...
 }));
 
+vi.mock('utils/recurrenceUtils', async () => {
+  const actual = await vi.importActual('utils/recurrenceUtils');
+  return {
+    ...actual,
+    validateRecurrenceInput: vi.fn(),
+    formatRecurrenceForApi: vi.fn(),
+  };
+});
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/screens/UserPortal/Events/Events.spec.tsx:56]: I’m intentionally keeping validateRecurrenceInput and formatRecurrenceForApi return values controlled per-test rather than setting defaults in beforeEach, so each test explicitly documents the intended behavior (valid vs invalid recurrence) and avoids hidden assumptions.

File: src/screens/UserPortal/Events/Events.spec.tsx [hunk 10]
```diff
@@ -1475,4 +1568,210 @@ describe('Testing Events Screen [User Portal]', () => {
...
+    }
+
+    fireEvent.submit(form);
+
+    await waitFor(() => {
+      expect(mockToast.error).toHaveBeenCalledWith('Invalid recurrence rule');
+    });
+
+    expect(formatRecurrenceForApi).not.toHaveBeenCalled();
+  });
```
[the remark this answers is not in the document]
Comment (rawadhossain) [reply src/screens/UserPortal/Events/Events.spec.tsx:1716]: The failure-path test already asserts that validation short-circuits the flow by checking toast.error is shown and formatRecurrenceForApi is not called. This verifies the behavior without coupling the test to Apollo internals or request inspection, which helps keep the test less brittle.

**Comment by rawadhossain** [comment]
@coderabbitai full review and reevaluate all resolved items and include the following points:
 
 1. Validate whether any updated test files:
    
    1. Cover all use cases
    2. Have valid tests
    3. Provide 100% test code coverage
    4. Require any additional tests
 2. Identify:
    
    1. Lines of code that are missing tests.
    2. Additional tests that may be required
    3. Any unnecessary files that have been submitted and ask that they be excluded.
 3. Validate that the PR:
    
    1. Meets the requirements of the linked issue.
    2. Complies with all previously suggested improvements and comments.
 
 If the submitted code requires any changes, even trivial ones, then make coderabbitai[bot] give the review a CHANGES_REQUESTED status. If no changes are required then give the PR an APPROVED status.
